### PR TITLE
Remove more Python 2.6 compatibility patches

### PR DESCRIPTION
**Pull request by astrofrog**
Some of the changes are easier to see if you ignore whitespace: https://github.com/astropy/astropy/pull/4486/files?w=0

cc @embray @eteq @taldcroft

### CODE REVIEW

File: docs/install.rst [hunk 1]
```diff
@@ -7,7 +7,7 @@ Requirements
 
 Astropy has the following strict requirements:
 
-- `Python <http://www.python.org/>`_ 2.6 (>=2.6.5), 2.7, 3.3, 3.4 or 3.5
+- `Python <http://www.python.org/>`_ 2.7, 3.3, 3.4 or 3.5
 
   - Prior to Astropy v1.0 Python 3.1 and 3.2 are also supported.
```
Comment (bsipocz) [inline docs/install.rst]: since 3.1 and 3.2 is mentioned here, maybe worth mentioning that 2.6 is also supported for astropy <v1.2, too

Comment (astrofrog) [inline docs/install.rst]: Good point - added now :)

File: astropy/tests/pytest_plugins.py [hunk 2]
```diff
@@ -35,10 +35,9 @@
 from ..utils.argparse import writeable_directory
 from ..utils.introspection import resolve_name
 
-# Needed for Python 2.6 compatibility
 try:
```
Comment (mhvk) [inline astropy/tests/pytest_plugins.py]: Can remove the whole try/except here, no?

**Comment by mhvk** [comment]
Modulo the small comment, the changes all look good. Have not checked for completeness, though...

File: astropy/tests/pytest_plugins.py (hunk 2 above)
Comment (astrofrog) [inline astropy/tests/pytest_plugins.py]: No - it's actually needed for Python 2.7 too, so I just removed the comment because it was wrong and added a comment on the except line :)

Comment (mhvk) [inline astropy/tests/pytest_plugins.py]: Ahh, teaches me to read properly before commenting....

**Comment by embray** [comment]
Something also needs to be done with this warning: https://github.com/astropy/astropy/blob/master/astropy/__init__.py#L161

I think it should stay there (and I think it should stay a warning and not an exception).  But maybe reworded to state that this version of Astropy definitely does not support Python 2.6.

**Comment by astrofrog** [comment]
@embray - do you think it should still be a deprecation warning?

**Comment by embray** [comment]
Maybe....?  I don't know.  The most important thing is that it's seen.  Though actually I wonder--with these changes does `import astropy` even work, in some minimal sense, on Python 2.6?

**Comment by astrofrog** [comment]
It doesn't even install actually:

```
mac-robitaille2:astropy tom$ python setup.py install
Traceback (most recent call last):
  File "setup.py", line 21, in <module>
    import astropy
  File "/Users/tom/Dropbox/Code/development/Astropy/astropy/astropy/__init__.py", line 115, in <module>
    from . import config as _config
  File "/Users/tom/Dropbox/Code/development/Astropy/astropy/astropy/config/__init__.py", line 11, in <module>
    from .paths import *
  File "/Users/tom/Dropbox/Code/development/Astropy/astropy/astropy/config/paths.py", line 9, in <module>
    from ..utils.decorators import wraps
  File "/Users/tom/Dropbox/Code/development/Astropy/astropy/astropy/utils/__init__.py", line 18, in <module>
    from .misc import *
  File "/Users/tom/Dropbox/Code/development/Astropy/astropy/astropy/utils/misc.py", line 24, in <module>
    from collections import defaultdict, OrderedDict
ImportError: cannot import name OrderedDict
```

so maybe I can just remove the warning?

**Comment by cdeil** [comment]
:+1: 

I'm getting this warning from the Astropy Sphinx build and the Sphinx build in affiliated packages (where I don't import `astropy.utils.compat.odict.OrderedDict` myself):
```
WARNING: AstropyDeprecationWarning: astropy.utils.compat.odict.OrderedDict is now deprecated - import OrderedDict from the collections module instead [astropy.utils.compat.odict]
```
I presume that is because that import is being used in Astropy itself?
@astrofrog - Is this being taken care of here?

**Comment by astrofrog** [comment]
@cdeil - can you point me to an example of a package where this happens? I just checked gammapy, and it is still importing OrderedDict from astropy.

**Comment by cdeil** [comment]
I see these warnings in Gammapy, even after removing the `astropy.utils.compat.odict.OrderedDict` imports ... see this branch:
https://github.com/gammapy/gammapy/pull/387

**Comment by astrofrog** [comment]
@cdeil - I don't see the warnings on that branch (and it looks like the sphinx builds are passing on Travis). Are you still seeing the issue?

Although I am seeing errors about a gammapy-extra repo, so is it possible there are references in there that need fixing?

**Comment by cdeil** [comment]
Yes, I'm still seeing the warnings.

For the Gammapy docs build, you need to do this:
```
git clone https://github.com/gammapy/gammapy-extra.git
export GAMMAPY_EXTRA=$PWD/gammapy-extra
```
before running the Sphinx build.

Related question: is there a way to configure `python setup.py build_sphinx` so that tracebacks are printed for warnings? That would be super-useful to locate where the warnings originate and remove all the guesswork.

**Comment by astrofrog** [comment]
@cdeil - try adding:

```
from astropy.tests.helper import enable_deprecations_as_exceptions
enable_deprecations_as_exceptions()
```

to the ``conf.py`` file. Otherwise you can also just edit the ``odict.py`` file in astropy to raise an exception.

I can't reproduce the warning here with your ``cdeil/hgps-browse`` branch.

**Comment by bsipocz** [comment]
@cdeil - Haven't tried it myself, but I wonder whether you get tracebacks when enabling raising the deprecation warnings as exceptions in ``gammapy/conftest.py``. It would only work locally (as sherpa still raises deprecation warnings), but maybe it can help trace these orderdict issues.

**Comment by cdeil** [comment]
Putting `enable_deprecations_as_exceptions()` in `docs/conf.py` or `gammapy/__init__.py` or the `gammapy/conftest.py` file has no effect.

The existing [sphinx-build -T option](http://www.sphinx-doc.org/en/stable/invocation.html#cmdoption-sphinx-build-T) is exactly what I want. Is there any way to invoke it via `python setup.py build_sphinx`?

**Comment by bsipocz** [comment]
@cdeil @astrofrog - I have the solution to the original problem. The warnings are coming from dependencies (again). Both wcsaxes and gwcs used the deprecated imports of OrderedDict. 

Somewhat related issue https://github.com/astropy/astropy/issues/4430. Maybe it would be also useful to (optionally) silence Warnings from dependencies, too?

**Comment by taldcroft** [comment]
I gave this a perusal of diffs and it looks good to me.  But I didn't test or look in detail.

**Comment by cdeil** [comment]
@bsipocz - Thanks for figuring this out.

My 2 cents: it would be very nice if:
- The `python setup.py build_sphinx` could pass arbitrary options to `sphinx-build`, just like is already possible with `-a` for `pytest`.
- Astropy and affiliated packages should run with the [sphinx-build -T](http://www.sphinx-doc.org/en/stable/invocation.html#cmdoption-sphinx-build-T) option by default. If warnings are printed, it's good to know where they come from, no?

Also: this PR looks good to me, so :+1:  to merge.
@astrofrog - Apologies for derailing it with my questions.

**Comment by astrofrog** [comment]
@cdeil - does ``-T`` actually apply to warnings as opposed to exceptions?

**Comment by bsipocz** [comment]
> The python setup.py build_sphinx could pass arbitrary options to sphinx-build, just like is already > possible with -a for pytest.

@cdeil - Could you open an issue for this in ``astropy-helpers`` so we don't forget that this need was raised? (Maybe it's trivial to implement/fix, but someone still needs to pick it up.)

**Comment by cdeil** [comment]
> does -T actually apply to warnings as opposed to exceptions?

Looks like it doesn't.

I tried adding
```
import warnings
warnings.filterwarnings("error")
```
in `docs/conf.py`, which works, but it exists on the first warning, which is usually not what I'm interested in.
Anyone know how to achieve the `-k` or `--keep-going` behaviour from make to continue the Sphinx build even if there's warnings / errors?


Looks like I already made this feature request for `build_sphinx` in 2014:
https://github.com/astropy/astropy-helpers/issues/83
:-)

**Comment by cdeil** [comment]
I did try find this:
http://stackoverflow.com/a/22376126
```
import traceback
import warnings
import sys

def warn_with_traceback(message, category, filename, lineno, file=None, line=None):
    traceback.print_stack()
    log = file if hasattr(file,'write') else sys.stderr
    log.write(warnings.formatwarning(message, category, filename, lineno, line))

warnings.showwarning = warn_with_traceback
```

If I put this in `docs/conf.py`, I get some useful extra info:
https://gist.github.com/cdeil/f69cfce09aebd75c9975

E.g. here the traceback points me exactly to the file / line in my code where I need to make a change:
https://gist.github.com/cdeil/f69cfce09aebd75c9975#file-gistfile1-txt-L1864

In other cases, e.g. when it comes from the plot directive, the traceback isn't very helpful, e.g.:
https://gist.github.com/cdeil/f69cfce09aebd75c9975#file-gistfile1-txt-L1956

Here it's pointing out a warning that originates in APLPy:
https://gist.github.com/cdeil/f69cfce09aebd75c9975#file-gistfile1-txt-L1074

I don't see the deprecation warnings from Astropy any more ... !?

Again, this is not pertinent to this PR, but might be useful background info for others (or my future self) trying to debug the origin of warnings.

**Comment by astrofrog** [comment]
I'll merge tomorrow if there are no objections.

**Comment by astrofrog** [comment]
@embray - I added back (and re-worded) the warning, and it's now at the top of ``__init__.py``

**Comment by eteq** [comment]
This looks good to me except one very minor wording change.  Will fix and then manual merge.
